atomicapp/providers/openshift.py

```diff
@@ -36,91 +35,205 @@ class OpenShiftProvider(Provider):
+            for artifact in objects:
+                namespace = self._get_namespace(artifact)
+                url = self._get_url(namespace, kind)
+
+                if self.dryrun:
+                    logger.info("DRY-RUN: %s", url)
+                    continue
+                res = requests.post(url, json=artifact, verify=self.ssl_verify)
+                if res.status_code == 201:
+                    logger.info(" %s sucessfully deployed.", res.content)
+                else:
+                    msg = "%s %s" % (res.status_code, res.content)
+                    logger.error(msg)
+                    # TODO: remove running components (issue: #428)
+                    raise ProviderFailedException(msg)
+
+    def undeploy(self):
+        logger.debug("starting undeploy")
+        # TODO: scale down replicationController before deleting deploymentConf
+        for kind, objects in self.openshift_artifacts.iteritems():
+            for artifact in objects:
+                namespace = self._get_namespace(artifact)
+
+                # get name from metadata so we know which object to be deleted
+                if "metadata" in artifact and \
+                        "name" in artifact["metadata"]:
+                    name = artifact["metadata"]["name"]
+                else:
+                    raise ProviderFailedException("Cannot undeploy. There is no"
+                                                  " name in artifacts metadata "
+                                                  "artifact=%s" % artifact)
+
+                url = self._get_url(namespace, kind, name)
+
+                if self.dryrun:
+                    logger.info("DRY-RUN: %s", url)
+                    continue
+
+                res = requests.delete(url, verify=self.ssl_verify)
+                if res.status_code == 200:
+                    logger.info(" %s sucessfully undeployed.", res.content)
+                else:
+                    msg = "%s %s" % (res.status_code, res.content)
+                    logger.error(msg)
+                    raise ProviderFailedException(msg)
+
+    def _process_artifacts(self):
+        """  parse artifact files
+
+        save parsed artifacts to self.openshift_artifacts
+        """
         for artifact in self.artifacts:
+            logger.debug("Procesesing artifact: %s", artifact)
             data = None
-            artifact_path = os.path.join(self.path, artifact)
-            with open(artifact_path, "r") as fp:
+            with open(os.path.join(self.path, artifact), "r") as fp:
                 data = anymarkup.parse(fp, force_types=None)
-            if "kind" in data:
-                if data["kind"].lower() == "template":
-                    logger.info("Processing template")
-                    artifact = self._processTemplate(artifact_path)
-                kube_order[data["kind"].lower()] = artifact
+                # kind has to be specified in artifact
```

Comment:
Any reason why the rest of this code can't be moved over out of the "with open()" block? We shouldn't need the fp around anymore after it has been read.